AlgebraInLean/Chapter02/Sheet04.lean

```diff
@@ -4,196 +4,209 @@ set_option linter.unusedTactic false
-arbitrary group G back to itself.
-
-As you have seen previously, in the context of Algebra, "group" is often omitted when discussing
-group endomorphisms. An endomorphism (and morphisms in general) can be defined among many
-different types of mathematical objects, but in AlgebraInLean it will always refer to a group
-endomorphism.
-
-Let's take a look at how this would be defined in Lean:
+Another special structure of morphism is the endomorphism, which is a morphism from an object to
+itself. As with other kinds of morphisms, the relevant structure is omitted when it can be inferred
+from context. Another kind of endomorphism you are accustomed to are square matrices in linear
+algebra (when interpreted as linear maps), which are vector space endomorphisms.
+
+From endomorphisms arises our final definition: the automorphism. An automorphism is a endomorphism
+that is also an isomorphism. We have already seen an example of this in the identity function, but
+there are more interesting options. You can think of automorphisms like a permutation of the
+elements of a group that preserves the group structure.
+
+Rather than defining these separately from `Homomorphism` in Lean, whether or not a particular
+homomorphism is an endomorphism can be directly determined from its type signature: merely check
+whether the domain and codomain are the same.
 -/
-def Endomorphism (φ : G → G) : Prop := Homomorphism φ
-
-/-
 
-A fairly simple definition, but important as we move on.
-
-Aside from group endomorphisms, a common example of an endomorphism is in linear algebra when
-considering some vector space V. f: V → V is an endomorphism on a vector space V, and we define
-_End(V)_ to be the set of all endomorphisms of V, which we know to be nonempty because of the
-existence of the endomorphism mapping some arbitrary vector v ↦ 0, and the identity mapping
-v ↦ v.
-
--/
+variable {G : Type*}
 
-
-
-/- An automorphism is defined to be an endomorphism that is also a bijection. You will recognize
-the following definition is similar to how we defined bijectivity in the first place. -/
-def Automorphism (φ : G → G) : Prop := Endomorphism φ ∧ Bijective φ
 /-
+One important automorphism is the inverse, but this is only an automorphism for abelian groups.
 
-You can think of it like a permutation from a group to itself, although it is important that
-this permutation respects the group structure. see more specifically what "respecting the group
-structure" looks like in the next chapter (keep an eye out for orders!).
-
+Hint: see `inv_anticomm`
 -/
-
-/- You may be able to intuitively discern that all automorphisms are also isomoprhisms.
-Lets do a brief exercise to prove this before jumping into some more complex examples. -/
-theorem aut_isomoprhism (φ : G → G) (h : Automorphism φ) : Isomorphism φ := by
+/-- -/
+theorem Group.inv_automorphism [AbelianGroup G] : Isomorphism (ι : G → G) := by
```

Comment:
This was not the proof that we had originally intended here -- Will recommended we add a short exercise proving that all automorphisms are also isomorphisms and I thought that was helpful

Reply:
It's not meant to be that (git can't really do a good job of what changes correspond to what when there are this many of them). That exercise doesn't really matter now that I'm just using `Isomorphism` for automorphisms (see response above). This proof is meant to be a more general version of the negation from the previous sheet (which had to be removed because it used Mathlib groups).